Sources/MessagingInApp/State/InAppMessageAction.swift

```diff
@@ -0,0 +1,76 @@
+    case clearMessageQueue
+    case loadMessage(message: Message, position: MessagePosition? = nil)
+    case embedMessage(message: Message, elementId: String)
+    case displayMessage(message: Message)
+    case dismissMessage(message: Message, shouldLog: Bool = true, viaCloseAction: Bool = true)
+    case reportError(message: String)
+    case engineAction(action: EngineAction)
+    case resetState
+
+    /// Represents an action that can be dispatched to InAppMessage store.
+    /// It only contains actions that are related to Gist engine view.
+    enum EngineAction: Equatable {
+        case tap(message: Message, route: String, name: String, action: String)
+        case messageLoadingFailed(message: Message)
+        case error(message: Message)
+    }
+
+    // swiftlint:disable cyclomatic_complexity
+    static func == (lhs: InAppMessageAction, rhs: InAppMessageAction) -> Bool {
+        switch (lhs, rhs) {
+        case (.initialize(let lhsSiteId, let lhsDataCenter, let lhsEnvironment), .initialize(let rhsSiteId, let rhsDataCenter, let rhsEnvironment)):
+            return lhsSiteId == rhsSiteId && lhsDataCenter == rhsDataCenter && lhsEnvironment == rhsEnvironment
+
+        case (.setPollingInterval(let lhsInterval), .setPollingInterval(let rhsInterval)):
+            return lhsInterval == rhsInterval
+
+        case (.setUserIdentifier(let lhsUser), .setUserIdentifier(let rhsUser)):
+            return lhsUser == rhsUser
+
+        case (.setPageRoute(let lhsRoute), .setPageRoute(let rhsRoute)):
+            return lhsRoute == rhsRoute
+
+        case (.processMessageQueue(let lhsMessages), .processMessageQueue(let rhsMessages)):
+            return lhsMessages == rhsMessages
+
+        case (.clearMessageQueue, .clearMessageQueue):
+            return true
+
+        case (.loadMessage(let lhsMessage, let lhsPosition), .loadMessage(let rhsMessage, let rhsPosition)):
+            return lhsMessage == rhsMessage && lhsPosition == rhsPosition
+
+        case (.embedMessage(let lhsMessage, let lhsElementId), .embedMessage(let rhsMessage, let rhsElementId)):
+            return lhsMessage == rhsMessage && lhsElementId == rhsElementId
+
+        case (.displayMessage(let lhsMessage), .displayMessage(let rhsMessage)):
+            return lhsMessage == rhsMessage
+
+        case (.dismissMessage(let lhsMessage, let lhsShouldLog, let lhsViaCloseAction), .dismissMessage(let rhsMessage, let rhsShouldLog, let rhsViaCloseAction)):
+            return lhsMessage == rhsMessage && lhsShouldLog == rhsShouldLog && lhsViaCloseAction == rhsViaCloseAction
+
+        case (.reportError(let lhsMessage), .reportError(let rhsMessage)):
+            return lhsMessage == rhsMessage
+
+        case (.engineAction(let lhsAction), .engineAction(let rhsAction)):
+            return lhsAction == rhsAction
+
+        case (.resetState, .resetState):
+            return true
+
+        default:
+            return false
+        }
+    }
```

Comment:
```suggestion
    // MARK: - Equatable Implementation
extension InAppMessageAction {
    static func == (lhs: InAppMessageAction, rhs: InAppMessageAction) -> Bool {
        switch (lhs, rhs) {
        case (.initialize(let l), .initialize(let r)): return l == r
        case (.setPollingInterval(let l), .setPollingInterval(let r)): return l == r
        case (.setUserIdentifier(let l), .setUserIdentifier(let r)): return l == r
        case (.setPageRoute(let l), .setPageRoute(let r)): return l == r
        case (.processMessageQueue(let l), .processMessageQueue(let r)): return l == r
        case (.clearMessageQueue, .clearMessageQueue): return true
        case (.loadMessage(let l), .loadMessage(let r)): return l == r
        case (.embedMessage(let l), .embedMessage(let r)): return l == r
        case (.displayMessage(let l), .displayMessage(let r)): return l == r
        case (.dismissMessage(let l), .dismissMessage(let r)): return l == r
        case (.reportError(let l), .reportError(let r)): return l == r
        case (.engineAction(let l), .engineAction(let r)): return l == r
        case (.resetState, .resetState): return true
        default: return false
        }
    }
}

// MARK: - Tuple Equatable Extensions
extension InAppMessageAction {
    private static func == (
        lhs: (siteId: String, dataCenter: String, environment: GistEnvironment),
        rhs: (siteId: String, dataCenter: String, environment: GistEnvironment)
    ) -> Bool {
        lhs.siteId == rhs.siteId && lhs.dataCenter == rhs.dataCenter && lhs.environment == rhs.environment
    }
    
    private static func == (
        lhs: (message: Message, position: MessagePosition?),
        rhs: (message: Message, position: MessagePosition?)
    ) -> Bool {
        lhs.message == rhs.message && lhs.position == rhs.position
    }
    
    private static func == (
        lhs: (message: Message, elementId: String),
        rhs: (message: Message, elementId: String)
    ) -> Bool {
        lhs.message == rhs.message && lhs.elementId == rhs.elementId
    }
    
    private static func == (
        lhs: (message: Message, shouldLog: Bool, viaCloseAction: Bool),
        rhs: (message: Message, shouldLog: Bool, viaCloseAction: Bool)
    ) -> Bool {
        lhs.message == rhs.message && lhs.shouldLog == rhs.shouldLog && lhs.viaCloseAction == rhs.viaCloseAction
    }
}
```

think we can simplify this and make it less error-prone by moving it to extension method and doing tuple comparisons wherever possible. 

e.g
in the case of  (.initialize(let l), .initialize(let r)): return l == r
Here, `l` and `r` are tuples containing the `siteId, dataCenter, and environment`. Swift automatically creates these tuples from the associated values of the enum case.
The comparison `l == r` calls the `custom == operator` we defined in another extension

So, when initialize is called, here's what happens:
- The == function in the main InAppMessageAction extension is called.
- It matches the .initialize case.
- It extracts the associated values into tuples l and r.
- It then calls l == r, which invokes our custom tuple comparison function.
- This function then compares each element of the tuple individually: siteId, dataCenter, and environment.

Reply:
Appreciate the write-up. However, honestly this seems too confusing to me. As it's hard to understand at first glance. Also, because the pairs/tuples used in the extension function are tightly bound and not very flexible, I think they are hard to read, maintain, and debug. So, I'm okay with moving forward with current approach, which might be error-prone but is clean and can be covered with tests to minimize the chances of breaking in future.

Reply:
> because the pairs/tuples used in the extension function are tightly bound and not very flexible, I think they are hard to read, maintain, and debug

I don't agree with this statement, as they are general equality operators that takes in tuples and if our action contains them, it applies to them too. 

I am okay with going forward this, subject to what tests gets added to ensure likelihood of errors when adding new cases or modifying existing ones.

Reply:
> they are general equality operators that take in tuples

No they are not, unfortunately, in my opinion. They are bound by number of parameters, their types, and the sequence. Consider the following example:

```swift
private static func == (
    lhs: (siteId: String, dataCenter: String, environment: GistEnvironment),
    rhs: (siteId: String, dataCenter: String, environment: GistEnvironment)
)
```

This probably will not work for any of the following:

```swift
private static func == (
    lhs: (environment: GistEnvironment, siteId: String, dataCenter: String),
    rhs: (environment: GistEnvironment, siteId: String, dataCenter: String)
)
``` 

OR

```swift
private static func == (
    lhs: (siteId: String, dataCenter: Region, environment: GistEnvironment),
    rhs: (siteId: String, dataCenter: Region, environment: GistEnvironment)
)
``` 

That's why I wouldn't call these general equality operators. Also, copying the code gives several compilation errors and tuple matching warnings (see screenshot below). So if I copied it correctly, I don't think it is working as expected 😅

<img width="1560" alt="Screenshot 2024-09-03 at 5 29 41 PM" src="https://github.com/user-attachments/assets/d05a51a8-68e5-41f5-a64d-a2ba1cfa7e45">

However, since we both agree on moving forward with current implementation, I think we should stick with it and continue.